### PR TITLE
feat: use external oc-patch-file-to-configmap-role to add dashboard json

### PR DESCRIPTION
**Pull request by darahayes**
This PR gives an example of how we could reuse an external role across different APBs to insert a dashboard json file into the desired config map.

This PR makes use of a role I created: https://github.com/aerogear/oc-patch-file-to-configmap-role 
This role takes 4 params:
- configmap name
- contents of a JSON file
- desired filename (inside the config map)
- namespace

And it inserts the file into the Config Map.

This role also gracefully handles cases where the specified Config Map does not exist (i.e. we are provisioning a service such as keycloak but we do not have metrics provisioned). It also handles cases where the `oc patch` command failed for some other reason.

**Note the keycloak-dashboard.json file is simply a placeholder that shows a blank panel**

## How the Role Gets Built into the APB

If you take a look in the `playbooks` directory, there is a `requirements.yml` file that specifies the dependency on https://github.com/aerogear/oc-patch-file-to-configmap-role.

In the Dockerfile there is an additional command:
```
RUN ansible-galaxy install -r /opt/apb/actions/requirements.yml -p /opt/ansible/roles
```

This command will install the roles specified in `requirements.yml` into the same directory as the APB roles (`/opt/ansible/roles`). What this means is that as part of `apb build` the required roles are built into the APB image.

Then in `playbooks/provision.yml` we can make use of the role and pass the appropriate parameters into it:

```
- role: oc-patch-file-to-configmap
    file: "{{ dashboard_file }}"
    filename: "{{ dashboard_filename }}"
    configmap: "{{ dashboards_configmap }}"
    namespace: "{{ namespace }}"
```

### CODE REVIEW

File: vars/main.yml [hunk 1]
```diff
@@ -10,3 +10,7 @@ keycloak_route_name: keycloak
 keycloak_secret_name: keycloak
 keycloak_pv_claim_name: keycloak-metrics
 postgres_secret_name: keycloak-postgres
+
+dashboards_configmap: grafana-dashboards-configmap
+dashboard_filename: keycloak-dashboard.json # name the file will have inside the config map
+dashboard_file: "{{ lookup('file','/opt/ansible/files/keycloak-dashboard.json') }}" # Contents of the dashboard file
```
Comment (david-martin) [inline vars/main.yml]: `dashboard_contents` or `dashboard_file_contents` might work better.
I was unsure from reading the definition in https://github.com/aerogear/oc-patch-file-to-configmap-role what the difference between `dashboard_filename` & `dashboard_file` were until I saw this example here.

**Comment by david-martin** [comment]
@pb82 Would like to get your thoughts on this

**Comment by pb82** [comment]
👀

File: vars/main.yml (hunk 1 above)
Comment (darahayes) [reply vars/main.yml]: Thanks @david-martin I agree, this name needs to be more explicit. Will update accordingly

**Comment by david-martin** [comment]
@darahayes Do you have any concerns with this approach? Are there other approaches you were considering?

**Comment by pb82** [comment]
I like this approach. Certainly better than repeating the oc patch and all the checks in every apb. Should we ever have more helper roles, we could rename the repo to something like `aerogear-asb-addons` and add them there.

**Comment by darahayes** [comment]
@david-martin The main concern/question is *how does this affect maintainability of parent APBs?*

In this example we're lucky enough that we run the provision role as one step and then we add the dashboard as another step. However, if we adopt this as a general approach to 'dependency management' for APBs I would like to find out:

Is it possible to use roles within roles? Having a fairly 'monolithic' *provision-my-service* type role seems to be pretty standard in APBS. What happens when we need to use some external role to perform a task in the middle of a larger provision role?  

I think a role within a role is a must because splitting up a larger role into a sequence of small roles would eventually lead to all-round headaches. Maybe I'm suffering from a slight lack of ansible expertise and 'a role within a role' might be really simple to do. I'd like to check that it's possible and then I'd be really happy with this approach.

The other consideration is figuring out is how to test these roles in isolation.

Lastly, it might be worth investigating using Ansible modules, written in Python. I've had a little bit of experience doing this before and the end result can be very tidy. I just figured it wasn't necessary in this case.

**Comment by david-martin** [comment]
@darahayes Understood. Are there examples of multiple roles being used for a provision. It might be worth reaching out to the ansible broker team on #asbroker.
I've seen plenty of examples of modules being used (which look nicer).
Is there a clear case where one approach would be chosen over the other?

**Comment by darahayes** [comment]
@david-martin I don't know officially when you should use one over the other.

My guess would be you build a module to do more low level stuff that would be difficult/verbose/very unmaintainable to do with straight up Ansible. For example, you might build a module to interface Ansible with an external API (Using a python client under the hood).

Whereas a role is much higher level thing that simply performs a task or a series of tasks.

We will probably have a good reason to build modules eventually but I don't think we do just yet. My reasoning behind building the 'oc-patch-file-to-configmap-role' as a role is as follows:

- Place a given file in the specified configmap (using bash), handle appropriate errors, do this in sequence. It is very task oriented. Seems like a good candidate for a role.
- You could write a python module that interfaces directly into the k8's API but you'd still need to repeat the same error handling across APBs.

### Next Steps

In `playbooks/provision.yml`, this PR declares roles like this:

```
- role: provision-keycloak-apb
     playbook_debug: false
- role: oc-patch-file-to-configmap
     file: "{{ dashboard_file }}"
     filename: "{{ dashboard_filename }}"
     configmap: "{{ dashboards_configmap }}"
     namespace: "{{ namespace }}"
```

This means the `provision-keycloak-apb` role is executed and then the `oc-patch-file-to-configmap` role is executed afterwards.

I think the next step is to check that it's possible to simply declare the `provision-keycloak-apb` role and then from somewhere within that role, we can make use of `oc-patch-file-to-configmap`. If this is possible, then the overall solution to external dependencies is golden IMO.

**Comment by darahayes** [comment]
I've pushed a new update. This change makes use of the `oc-patch-file-to-configmap` role **from within** the `keycloak-provision-role`.

Specifically in `roles/provision-keycloak-apb/tasks/main.yml` we make use of a `include_role` feature in Ansible:

```yaml
- include_role:
    name: oc-patch-file-to-configmap
  vars:
    file_contents: "{{ dashboard_file_contents }}"
    filename: "{{ dashboard_filename }}"
    configmap: "{{ dashboards_configmap }}"
    namespace: "{{ namespace }}"
```

Given that this is possible, I'm really happy with this approach. What do you think @pb82 @david-martin ?

**Comment by david-martin** [comment]
@darahayes I'm good with this. Makes it cleaner than original approach.

**Pull request merged**